runtime/sema/type.go

```diff
@@ -5502,6 +5506,19 @@ func checkSubTypeWithoutEquality(subType Type, superType Type) bool {
 			//
 			// The holder of the reference may not gain more permissions or knowledge.
 
+			return false
+
+		case *InterfaceType:
+			switch typedInnerSubType := typedSubType.Type.(type) {
+			case *CompositeType:
+				return typedInnerSubType.ExplicitInterfaceConformanceSet().Contains(typedInnerSuperType)
+			// An interface type is a supertype of a restricted type if the restricted set contains
+			// that explicit interface type. Once interfaces can conform to interfaces, this should instead
+			// check that at least one value in the restriction set is a subtype of the interface supertype
+			case *RestrictedType:
+				return typedInnerSubType.RestrictionSet().Contains(typedInnerSuperType)
+			}
```

Comment:
This is needed to make `&{I}` a subtype of `&I`, and thus to allow an attachment for `I` to be accessed on a reference restricted to `{I}`

Reply:
Maybe promote the `I` to a restricted type instead.

Reply:
After some thought, requiring manual conversion from an interface to a singleton restricted set every time the base type of an attachment is used is pretty fragile; forgetting to perform the conversion in some case could easily cause a missed error. I'm pretty confident this logic is correct and feel like adding support here is better than requiring all potential inputs to this function to be converted.

Reply:
> The implementation ignores the restricted type (T in &T{Us}), does it need to be checked too?

We are deliberately excluding this check: 

```
struct R: I, J {}
struct interface I {}
struct interface J {}
attachment A for I {}
pub fun foo(r: R{J}) {
	r[A]
}
```

We would like the above code to fail to typecheck; although `R` is a subtype of `I`, when it is restricted to `J` we don't want people to be able to access or add attachments to it that are only intended to work with `I` values.